8387875: [lworld] Provide value-class-preview.md file (#2633)
openjdk/valhalla: doc/value-class-preview.md

% Migration of JDK Classes to Value Classes

## Introduction

> The "Value Objects" JEP introduces value objects and migrates suitable classes

Reviewer: Rewording suggestion to avoid the need to say "JEP"; something like: "Value objects and value classes are a preview feature that migrates selected platform classes to value classes. This mean when ..."

> Value Objects feature introduces value objects and migrates suitable classes to

Reviewer 2: Either "The Value Objects feature introduces ..." or "Value Objects is a feature that introduces ..."

to value classes. This means that when preview features are enabled, a different

> classpath is available to Java programs.

Reviewer: Nit: is it a _class_path or a _module_path or does that detail not matter here?

To accomplish this, the JDK uses *preview-specific* files in `META-INF/preview`,

> which overrides files of the same name in the regular classpath. For example,

Reviewer 2: "classpath"? Isn't this the modulepath these days?

Author: I use the term "classpath" for where the resources like class files are discovered. I don't know if we ever use the term "modulepath".

Reviewer: Both the `java` and `javac` commands have "module path" options, along with "class path" options.

Author: Is there a better term to refer to directories or locations where such resources? Maybe something like "overrides files of the same name in the regular locations"?

Reviewer 3: "To accomplish this, the JDK uses *preview-specific* files in `META-INF/preview` that will be used instead of the original file when running in preview mode"?

Author: Done. Please review again.

`META-INF/preview/java/lang/Integer.class` overrides `java/lang/Integer.class`.

The JDK generates preview-specific source files (they may use preview language
features), compiles class files from them, and distributes these class files in
`META-INF/preview`.

## The Build Process

### The Custom Handling for the Value Objects JEP

The Value Objects JEP requires a few select classes in the `java.base` module
to become value classes when preview features are enabled.

> The build of `java.base` module first create the source code of those value

Reviewer 3: ```suggestion
The build of `java.base` module first creates the source code of those value
```

classes, done in
[`GensrcValueClasses.gmk`](../make/modules/java.base/gensrc/GensrcValueClasses.gmk).

1. A hardcoded list of files are selected for generation.

2. For each file, search for any occurrences of `/*value*/ class` or
   `/*value*/ record`, and replace with `value class` or `value record`.

3. The replaced, preview-specific source files are located in
   `support/gensrc-valueclasses/java.base/`.

4. The general preview source to binary build pipeline recognizes this

>    directory as where `java.base` places its preview-specific source files.

Reviewer 3: @liach This text is sort of ambiguous about whether the original files still remain in place, I think it should be spelled out that the original remains in place, and that the value-capable version of the source is a transformed copy.

build system into a fully automated pipeline handling all modules and all
outcome images.

1. The `GENERATED_PREVIEW_SUBDIRS` variable in [`make/common/Modules.gmk`](../make/common/Modules.gmk)
   indicates where the source files are found.

2. For each module that has preview-specific source files, a goal is created
   to compile these source files into class files.

3. The class files from each of these tasks reside in `support/preview/<module>`
   for each module.

4. These preview-specific class files and other resources are copied to the
   `META-INF/preview` directory of the regular output directory.

5. At run-time, jimage will pick up the preview-specific overrides from
   `META-INF/preview`  when preview features are enabled.

6. The interim `javac` used by the build system cannot pickup the
   preview-specific overrides; they must be supplied explicitly with this
   javac flag for every single module where overrides are significant:

   ```
   --patch-module <module>=$(SUPPORT_OUTPUTDIR)/preview/<module>
   ```

   See [`BuildMicroBenchmarks.gmk`](../make/test/BuildMicrobenchmark.gmk) for an example.

> 

Reviewer: I recommend explicitly stating up front "Different class files are used at runtime when preview is enabled."

I believe that is a new-with-JEP-401 implementation technique for JEPs.

% Migration of JDK Classes to Value Classes

## Introduction

The Value Objects feature introduces value objects and migrates suitable classes
to value classes. This means that when preview features are enabled, different
class files are used for the migrated classes in the Java class library.

> To accomplish this, the JDK uses *preview-specific* files in `META-INF/preview`,

Reviewer: To clarify,
"To accomplish this, a built JDK uses..."

That is, this section is discussing the artifact people run, not the source structure, etc.

   located in `support/gensrc-valueclasses/java.base/`.  The regular source
   files remain unchanged in their original locations.

4. The general preview source to binary build pipeline recognizes the
   `support/gensrc-valueclasses/java.base/` directory as where the `java.base`
   module places its preview-specific source files.

### The General Preview Source to Binary Pipeline

Once the preview-specific source files are ready, they are picked up by the
build system into a fully automated pipeline handling all modules and all
outcome images.

1. The `GENERATED_PREVIEW_SUBDIRS` variable in [`make/common/Modules.gmk`](../make/common/Modules.gmk)
   indicates where the source files are found.

2. For each module that has preview-specific source files, a goal is created
   to compile these source files into class files.

3. The class files from each of these tasks reside in `support/preview/<module>`
   for each module.

4. These preview-specific class files and other resources are copied to the
   `META-INF/preview` directory of the regular output directory.

5. At run-time, jimage will pick up the preview-specific overrides from

>    `META-INF/preview` only when features are enabled.

Reviewer 4: when _preview_ features

>    `META-INF/preview`  when preview features are enabled.

Reviewer 3: ```suggestion
5. At run-time, jimage will pick up the preview-specific overrides from
   `META-INF/preview` only when preview features are enabled.
```

6. The interim `javac` used by the build system cannot pickup the

>    preview-specific overrides; they must be supplied explicitly with this

Reviewer 3: ```suggestion
   preview-specific overrides; they must be supplied explicitly with the following
```

   following javac flag for every single module where overrides are significant:

   ```
   --patch-module <module>=$(SUPPORT_OUTPUTDIR)/preview/<module>
   ```

   See [`BuildMicroBenchmarks.gmk`](../make/test/BuildMicrobenchmark.gmk) for an example.

### Non-Goals

The Value Objects JEP only plans to introduce value classes that are:

1. In the `java.base` module.

   There's no plan to migrate other classes in other modules.

2. Migrated from existing classes.

   These classes are available as identity classes when preview features are
   disabled. There's no plan to introduce completely new value classes.

> Support for other value classes require significant changes to the build system.

Reviewer 4: _would_ require

## Testing

In addition to tests that require preview features to be enabled, tests that do
not depend on preview features wish to run with preview features enabled to
ensure compatibility:

1. Some tests wish to run against the Java SE class library with value classes.

   The jtreg tests may be run with `JTREG=VM_OPTIONS=--enable-preview`.

2. Some tests wish to run against their own classes migrated to value classes.

   The jtreg [VALUE_CLASS_PLUGIN](testing.html#VALUE_CLASS_PLUGIN) allows tests
   to migrate their own classes to value classes when running with the plugin.

## Wrapper Class Caches

Currently, wrapper class caches are retained even when preview features are
enabled. They are created with `ValueClass.newReferenceArray`, which creates
a non-flat array storing object references, allowing a value object wrapping
a primitive value to be available quickly without redundant allocations and

> associated performance regressions. Their existence have no semantic meaning,

Reviewer 4: _has_ no semantic meaning

% Migration of JDK Classes to Value Classes

## Introduction

The "Value Objects" JEP introduces value objects and migrates suitable classes
to value classes. This means that when preview features are enabled, a different
classpath is available to Java programs. To accomplish this, the JDK has a

> custom build and deployment process to allow access to different classpaths.

Reviewer: I think this intro should explicit words to the effect of "there are separate class files for migrated value classes with preview enabled and with preview enabled. First new source files are generated from the original files to be migrated and then these new source files are compiled into separate class files are put into a distinct class path, etc."

(The salient point here for core libs maintainers include there is a new source-based transform and, as explained later in the note, understanding the limited functionality of the transform. Other technically possible ways to do this transform would include class file re-writing, possibly at startup, etc.)

Reviewer 3: I think Joe makes a good point here

Author: Added a somewhat better overview for this.

## The Build Process

The build system has a fully automated pipeline from preview-only source files
for each module to preview-ready image binaries.  The source files are currently
hardcoded in `GENERATED_PREVIEW_SUBDIRS` variable in
[`make/common/Modules.gmk`](../make/common/Modules.gmk).

### Hardcoding for Value Objects JEP

For the Value Objects JEP, for each module, the preview source files are in
`$(SUPPORT_OUTPUTDIR)/gensrc-valueclasses/<module>/` directory. However, only
the module `java.base` actually makes use of this directory, created by
[`GensrcValueClasses.gmk`](../make/modules/java.base/gensrc/GensrcValueClasses.gmk).
This processor simply looks for occurrences of `/*value*/ class` and
`/*value*/ record` in the hardcoded list of source files, and generate preview
source files where these occurrences are replaced by `value class` and
`value record`, respectively.

> 

Reviewer: So, with the current system, consequently, any needed testing of "is preview enabled" must be done at runtime rather than via a compile-time macro?
(The other little code gen systems used in the JDK have a more expressive set of primitives. In my estimation, the simple system here is good for the purpose it is being used for, but it is helpful to document it here.)

Author: Yes. I don't understand how a compile-time macro can work - we have one runtime that must support both preview-on and preview-off runs.

Reviewer: In the space of generate-different-source-code-from-a-template solutions, it would be possible to have code conditionally emitted for one case but not the other. The solution used for JEP 401 currently just uncomments the `value` modifier.

Author: We tried to use annotation processors and discovered that it just introduced unnecessary hassles and complexities beyond what is required for JEP 401. I have described the JEP 401 process here; we can tweak this document if we do tweak this pipeline in the future.

Reviewer: I understand some of the history for the transitory use of annotation processors, but the readers of this document might have not that history (and might not care about it either ;-) ). 

Given that the general shape of the solution is "we have generating preview-specific versions of source files which get compiled into preview-specific class files," the current mechanism we are choosing to do the generation has very limited functionality, which is sufficient for the purposes at hand. However, that mechanism is less capable than several analagous mechanisms elsewhere in the JDK build.

> The Value Objects JEP does not intend to introduce value classes that are not

Reviewer 3: Too many negations in this sentence

Author: Done, converted this to a list of prerequisites.

migrated from a class that is available when preview features are disabled, nor
does it intend to introduce migrations from modules other than `java.base`. The
preview source file setup would require substantial changes to accomplish these
goals.

### Preview Classes in Interim `javac`

The preview classpath files are not available to the interim `javac`, most
commonly used in `SetupJavaCompilation` call. For each module that ships

> preview-specific binaries, the following javac flag is required for a

Reviewer 3: What's the definition of "preview-specific binaries" here? If we mean "value object-capable binaries" I think we should be clear about that.

Author: Now defined at top about how "preview-specific" stuff overrides regular stuff.

classpath is available to Java programs.

To accomplish this, the JDK uses *preview-specific* files in `META-INF/preview`,
which overrides files of the same name in the regular classpath. For example,
`META-INF/preview/java/lang/Integer.class` overrides `java/lang/Integer.class`.

The JDK generates preview-specific source files (they may use preview language
features), compiles class files from them, and distributes these class files in
`META-INF/preview`.

## The Build Process

### The Custom Handling for the Value Objects JEP

The Value Objects JEP requires a few select classes in the `java.base` module
to become value classes in the classpath when preview features are enabled.

The build of `java.base` module first create the source code of those value
classes, done in
[`GensrcValueClasses.gmk`](../make/modules/java.base/gensrc/GensrcValueClasses.gmk).

1. A hardcoded list of files are selected for generation.

2. For each file, search for any occurrences of `/*value*/ class` or
   `/*value*/ record`, and replace with `value class` or `value record`.

3. The replaced, preview-specific source files are located in
   `support/gensrc-valueclasses/java.base/`.

> 4. The general preview source to binary pipeline recognizes this

Reviewer 3: *build pipeline*

build system into a fully automated pipeline handling all modules and all
outcome images.

1. The `GENERATED_PREVIEW_SUBDIRS` variable in [`make/common/Modules.gmk`](../make/common/Modules.gmk)
   indicates where the source files are found.

2. For each module that has preview-specific source files, a goal is created
   to compile these source files into class files.

3. The class files from each of these tasks reside in `support/preview/<module>`
   for each module.

4. These preview-specific class files and other resources are copied to the
   `META-INF/preview` directory of the regular output directory.

5. At run-time, jimage will pick up the preview-specific overrides from
   `META-INF/preview`  when preview features are enabled.

6. The interim `javac` used by the build system cannot pickup the
   preview-specific overrides; they must be supplied explicitly with this
   javac flag for every single module where overrides are significant:

   ```
   --patch-module <module>=$(SUPPORT_OUTPUTDIR)/preview/<module>
   ```

   See [`BuildMicroBenchmarks.gmk`](../make/test/BuildMicrobenchmark.gmk) for an example.

### Non-Goals

> The Value Objects JEP only plans to introduce a value class that:

Reviewer 3: >The Value Objects JEP only plans to introduce a value class that:

This doesn't read right to me, perhaps we mean something more akin to "The Value Objects JEP only migrates a select few platform types that:"
